# 从恐慌中恢复

> 原文：<https://golangbyexample.com/recover-panic-goroutine-go/>

# **概述**

有两个从恐慌中恢复的案例

*   **恢复**功能的方式与**恐慌**相同

*   **恢复**功能在不同的戈罗廷作为**恐慌**

而在第一种情况下，它会从恐慌中恢复过来。但是关于`recover()`函数需要注意的一点是，它只能恢复发生在同一个 goroutine 中的恐慌。如果恐慌发生在不同的领域，而复苏发生在不同的领域。

让我们看一个两者的例子

# ****`recover()`函数与**** `recover()`函数相同

```go
package main

import "fmt"

func main() {
    a := []string{"a", "b"}
    go checkAndPrintWithRecover(a, 2)
    fmt.Println("Exiting normally")
}

func checkAndPrintWithRecover(a []string, index int) {
    defer handleOutOfBounds()
    checkAndPrint(a, 2)
}

func checkAndPrint(a []string, index int) {
    if index > (len(a) - 1) {
        panic("Out of bound access for slice")
    }
    fmt.Println(a[index])
}

func handleOutOfBounds() {
    if r := recover(); r != nil {
        fmt.Println("Recovering from panic:", r)
    }
}
```

**输出**

```go
Exiting normally
Recovering from panic: Out of bound access for slice
```

在上面的程序中**恢复**和**恐慌**位于同一条线上，因此从输出可以看到它能够从恐慌中恢复。

# **`recover()`函数在不同的戈罗廷一样恐慌**

正如在这种情况下提到的，从恐慌中恢复是不可能的。

```go
package main
import "fmt"
func main() {
    a := []string{"a", "b"}
    checkAndPrintWithRecover(a, 2)
    fmt.Println("Exiting normally")
}
func checkAndPrintWithRecover(a []string, index int) {
    defer handleOutOfBounds()
    go checkAndPrint(a, 2)
}
func checkAndPrint(a []string, index int) {
    if index > (len(a) - 1) {
        panic("Out of bound access for slice")
    }
    fmt.Println(a[index])
}
func handleOutOfBounds() {
    if r := recover(); r != nil {
        fmt.Println("Recovering from panic:", r)
    }
}
```

**输出**

```go
Exiting normally
panic: Out of bound access for slice

goroutine 18 [running]:
main.checkAndPrint(0xc0000a6020, 0x2, 0x2, 0x2)
        /Users/slohia/go/src/github.com/golang-examples/articles/tutorial/panicRecover/goroutine/main.go:19 +0xe2
created by main.checkAndPrintWithRecover
        /Users/slohia/go/src/github.com/golang-examples/articles/tutorial/panicRecover/goroutine/main.go:14 +0x82
exit status 2
```

在上面的程序中，我们在 goro tine 中有`checkAndPrint`，它在 goro tine 中引发**恐慌**。**恢复**功能在通话中。从输出中可以看出，它不会停止**恐慌**，因此您会看到 a